tests/30rooms/20typing.pl

```diff
@@ -109,27 +106,22 @@ sub matrix_typing
    do => sub {
       my ( $typinguser, $local_user, $room_id ) = @_;
 
-      matrix_typing( $typinguser, $room_id, typing => 0 )->then( sub {
+      matrix_typing( $typinguser, $room_id, typing => JSON::false )->then( sub {
          Future->needs_all( map {
             my $recvuser = $_;
 
-            await_event_for( $recvuser, filter => sub {
-               my ( $event ) = @_;
-
-               return unless $event->{type} eq "m.typing";
-
-               assert_json_keys( $event, qw( type room_id content ));
-               assert_json_keys( my $content = $event->{content}, qw( user_ids ));
-
-               return unless $event->{room_id} eq $room_id;
-
-               assert_json_list( my $users = $content->{user_ids} );
+            await_sync_ephemeral_contains($recvuser, $room_id,
+               check => sub {
+                  my ( $event ) = @_;
+                  return unless $event->{type} eq "m.typing";
 
-               scalar @$users and
-                  die "Expected 0 members to be typing";
+                  assert_json_keys( $event, qw( type content ));
+                  assert_json_keys( my $content = $event->{content}, qw( user_ids ));
+                  assert_json_list( my $users = $content->{user_ids} );
 
-               return 1;
-            })
+                  return scalar @$users == 0;
```

Comment:
if this doesn't match, shouldn't we raise an error rather than continue waiting for another typing event?

Reply:
It originally raised an error which caused Synapse in worker mode to fail, presumably because in worker mode it might redundantly send the typing:true event or something. Waiting a bit longer for the typing:false event makes this test pass in worker mode.

Reply:
It's surprising that this used to work reliably, but if what you say is correct...

The problem is that tests that just time out rather than failing when the code isn't doing the right thing are something of an antipattern (mostly: you can't tell if the server was slow or there is an actual bug). We already have too many tests in sytest that do this, and I am keen that we don't add any more.

It's hard to completely eliminate this, but to make it better, I suggest you refactor it so that you check each returned typing event, and more than one (?) event with the wrong number of users.

Most importantly though, you need to `log_if_fail` for each "rejected" typing event, saying why it is rejected. One way to do this is to wrap the whole thing in a `retry_until_success` loop with an `on_fail` handler (see https://github.com/matrix-org/sytest/blob/master/tests/30rooms/13guestaccess.pl#L414-L418 for example), but there are plenty of other solutions.

Reply:
I've made the following changes:
 - We log each typing event received. We also log why the event is being ignored if it doesn't meet the criteria (number of users).
 - If we encounter more than 1 typing event with the wrong number of users, we bail out the test.

Produces output like: (ignore the fact that there are no users typing, I had to invert the checks to get this output)
```
not ok 3 (expected fail) Typing can be explicitly stopped # TODO expected fail
# Started: 2020-06-23 11:10:43.737
# Ended: 2020-06-23 11:10:43.756
# seen too many typing events with typing users at tests/30rooms/20typing.pl line 130.
# 0.017327: Received typing event: 
# {
#   content => { user_ids => [] },
#   event_id => "",
#   origin_server_ts => JSON::number(0),
#   sender => "",
#   type => "m.typing",
# }
# 0.017624: rejecting event because want zero users typing, but there are some
```